docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,777 @@
+
+```golang
+func ConnOpenInit(local ConnectionParty, remote ConnectionParty)
+{
+  // Create our own end of the connection.
+  connectionEnd.state = INIT
+  connectionEnd.local = local
+  connectionEnd.remote = remote
+  connectionEnd.version = getCompatibleVersions()
+
+  // Validity check. (Explained below in helper functions).
+  abortTransactionUnless(validate(connectionEnd))
+
+  // Uniqueness check. (Explained below in helper functions).
+  abortTransactionUnless(unique(connectionEnd))
+
+  // Now we save this connectionE in our local store.
+  // Since this connectionEnd is in state INIT, the relayer will continue with
+  // next steps of the connection handshake protocol.
+  setConnection(local.connectionIdentifier, connectionEnd)
+}
+```
+
+Preconditions:
+  - the connection (obtained from the input `local` and `remote` arguments) is valid and unique
+
+Postconditions:
+  - the module executing this handler stores a `ConnectionEnd` (matching with the input arguments `local` and `remote`) in state `INIT`
+  - function produces no output; may abort if preconditions not met
+
+Implementation remarks:
+  - to ensure termination of the ICS3 protocol, this handler must execute at least once (either one of the chains or both)
+
+
+###### ConnOpenTry
+
+This is step 2.b.
+
+```golang
+func ConnOpenTry(local ConnectionParty,
+                 remote ConnectionParty,
+                 remoteVersions string[],
+                 remoteStateProof Proof,
+                 remoteClientProof Proof)
+{
+  // Create our end of the connection.
+  connectionEnd.state = TRYOPEN
+  connectionEnd.local = local
+  connectionEnd.remote = remote
+  connectionEnd.version = pickVersion(remoteVersions)
+
+  // Validity check.
+  abortTransactionUnless(validate(connectionEnd))
+
+  // Uniqueness check.
+  abortTransactionUnless(unique(connectionEnd))
+
+  // `verifyProofs` does a complete verification of proofs.
+  // See definition below in helper functions.
+  abortTransactionUnless(
+    verifyProofs(remoteStateProof, remoteClientProof, connectionEnd))
+
+  // Handler is done: store the connectionEnd and let the relayer handle
+  // the next step.
+  setConnection(local.connectionIdentifier, connectionEnd)
+}
+```
+
+Preconditions:
+  - the connection (obtained from inputs `local` and `remote`) is valid and unique
+  - the module running this handler supports at least one of the versions supplied in the input list `remoteVersions`
+  - the two proofs `remoteStateProof` and `remoteClientProof` are syntactically and semantically correct; this is necessary in connection to properties [ICS3-Proto-3-ConnectionIntegrity] and [ICS3-Proto-4-StateConsistency]
+
+Postconditions:
+  - the module executing this handler stores a `ConnectionEnd` (matching with the arguments `local` and `remote`) in state `TRYOPEN`
+  - no output; may abort if preconditions not met
+
+Implementation remarks:
+  - the `validate` check should also check the field `connectionEnd.version`.
```

Comment:
not sure what this check entails. I assume `validate` validates all fields.

Reply:
There is a full description of this function further down in the spec:

> ###### `validate(ConnectionEnd c)` 
> This checks that the object `connectionEnd.local.connectionIdentifier` is valid.
Similar in functionality with `validateConnectionIdentifier` from [ICS 003]; additionally, this function should also validate the `version` field.